prefect_dbt/cloud/jobs.py

```diff
@@ -318,9 +326,182 @@ def my_flow():
+            except Exception:
+                pass
+        else:
+            raise DbtCloudJobRunFailed(f"Triggered job run with ID: {run_id} failed.")
     else:
         raise RuntimeError(
             f"Triggered job run with ID: {run_id} ended with unexpected"
             f"status {final_run_status.value}."
         )
+
+
+async def _build_trigger_job_run_options(
+    dbt_cloud_credentials: DbtCloudCredentials,
+    trigger_job_run_options: TriggerJobRunOptions,
+    run_id: str,
+    run_info: Dict[str, Any],
+):
+    """
+    Compiles a list of steps (commands) to retry, then either build trigger job
+    run options from scratch if it does not exist, else overrides the existing.
+    """
+    steps_override = []
+    for run_step in run_info["run_steps"]:
+        status = run_step["status_humanized"].lower()
+        # Skipping cloning, profile setup, and dbt deps - always the first three
+        # steps in any run, and note, index starts at 1 instead of 0
+        if run_step["index"] <= 3 or status == "success":
+            continue
+        # get dbt build from "Invoke dbt with `dbt build`"
+        command = run_step["name"].partition("`")[2].partition("`")[0]
+        sub_command = shlex.split(command)[1]
+
+        is_run_command = command in ("build", "run", "test", "seed", "snapshot")
+        is_not_success = status in ("error", "skipped", "cancelled")
+        is_skipped = status == "skipped"
+        if (not is_run_command and is_not_success) or (is_run_command and is_skipped):
+            steps_override.append(command)
+        # errors and failures are when we need to inspect to figure
+        # out the point of failure
+        else:
+            # get the run results scoped to the step which had an error
+            # an error here indicates that either:
+            # 1) the fail-fast flag was set, in which case
+            #    the run_results.json file was never created; or
+            # 2) there was a problem on dbt Cloud's side saving
+            #    this artifact
+            try:
+                run_artifact_future = await get_dbt_cloud_run_artifact.submit(
+                    dbt_cloud_credentials=dbt_cloud_credentials,
+                    run_id=run_id,
+                    path="run_results.json",
+                    step=run_step["index"],
+                )
+                run_artifact = await run_artifact_future.result()
+                run_results = run_artifact["results"]
+            except DbtCloudGetRunArtifactFailed:
+                steps_override.append(command)
+            else:
+                run_nodes = " ".join(
+                    run_result["unique_id"].split(".")[2]
+                    for run_result in run_results
+                    # "fail" here instead of "cancelled"
+                    if run_result["status"] in ("error", "skipped", "fail")
+                )
+                modified_command = f"dbt {sub_command} --select {run_nodes}"
```

Comment:
This modified command would still omit variables that a step could possibly have.  For example, a step could be:
```
dbt build --vars '{"env": "prod"}'
```
The key: value pairs for `vars` are user-defined.  And, generally speaking, these are pretty important, as they could define where to grab data from, where to build data to, define date ranges, etc.

You can see I've defined the different args available [here](https://github.com/dpguthrie/dbtc/blob/ae83d5dd8ec77fffd40ddb95b19e0f262e01d7d6/dbtc/client/cloud/base.py#L26-L39).  And you can see how I'm creating a modified command after parsing for each of those args [here](https://github.com/dpguthrie/dbtc/blob/ae83d5dd8ec77fffd40ddb95b19e0f262e01d7d6/dbtc/client/cloud/base.py#L1140-L1146)

Let me know if that's unclear.

Reply:
Thanks again for the context and example! [I took another route](https://github.com/PrefectHQ/prefect-dbt/pull/81/commits/78bc769c744cdbdfab557aaa3db36f234d81c86e) of simply partitioning by the sub_command  (like build) and rebuilding the command with select.

However, I'm curious, if there's a difference between:
1. `dbt build --vars '{"env": "prod"}' --select node1 node2`
2. `dbt build --select node1 node2 --vars '{"env": "prod"}'`

Would 1 work? If so, do I need to partition, or simply just append the select arg?

I really appreciate your insights!